qiskit_experiments/framework/base_experiment.py

```diff
@@ -167,29 +144,114 @@ def _initialize_experiment_data(
 
         return experiment_data._copy_metadata()
 
-    def run_analysis(self, experiment_data, **options) -> ExperimentData:
+    def _pre_transpile_hook(self, backend: Backend):
+        """An extra subroutine executed before transpilation.
+
+        Args:
+            backend: Target backend.
+        """
+        pass
+
+    # pylint: disable = unused-argument
+    def _post_transpile_hook(
+        self, circuits: List[QuantumCircuit], backend: Backend
+    ) -> List[QuantumCircuit]:
+        """An extra subroutine executed after transpilation.
+
+        Args:
+            circuits: List of transpiled circuits.
+            backend: Target backend.
+
+        Returns:
+            List of circuits to execute.
+        """
+        return circuits
+
+    def run_transpile(self, backend: Backend, **options) -> List[QuantumCircuit]:
```

Comment:
This causes a problem in `CompositeExperiment.run_transpile`. We need to apply transpile options and pre/post method for each experiment, thus `circuits` here will be `List[List[QuantumCircuits]]` for `CompositeExperiment` while it will be `List[QuantumCircuit]` for `BaseExperiment`. This is why circuit generation is not done in the run method, because we cannot combine un-transpiled circuits as currently implemented in the composite experiments (current logic doesn't support mixture of different experiments).

Another advantage would be https://github.com/Qiskit/qiskit-experiments/pull/380#issuecomment-915371441. User can easily check what will be executed. If we assume circuit is always generated by `run` method, the `run_transpile` should be a protected method.

Reply:
According to your comment, probably `run_transpile` is not correct name for this function. Something like `run_circuit_generation` would make more sense?